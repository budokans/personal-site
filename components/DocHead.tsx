import Head from "next/head";
import { ReactElement } from "react";
import { SiteMetadata } from "types";

interface DocHeadProps {
  readonly metadata: SiteMetadata;
}

export const DocHead = ({ metadata }: DocHeadProps): ReactElement => (
  <Head>
    <title>{metadata.title}</title>
    <meta name="description" content={metadata.description} />

    <meta property="og:title" content={metadata.title} />
    <meta property="og:description" content={metadata.description} />
    <meta property="og:image" content={`${metadata.canonical}/og.png`} />
    <meta property="og:locale" content="en_US" />
    <meta property="og:type" content="website" />
    <meta property="og:site_name" content="Steven Webster" />

    <meta name="twitter:card" content="summary_large_image" />
    <meta name="twitter:title" content={metadata.title} />
    <meta name="twitter:image:src" content={`${metadata.canonical}/og.png`} />
    <meta name="twitter:description" content={metadata.description} />
    <meta name="twitter:url" content={metadata.canonical} />

    <link rel="icon" type="image/png" sizes="32x32" href="/favicon-32x32.png" />
    <link rel="icon" type="image/png" sizes="16x16" href="/favicon-16x16.png" />
    <link rel="apple-touch-icon" sizes="180x180" href="/apple-touch-icon.png" />
    <link rel="manifest" href="/site.webmanifest" />
    <link rel="mask-icon" href="/safari-pinned-tab.svg" color="#5bbad5" />
    <meta name="theme-color" content="#ffffff" />
  </Head>
);
