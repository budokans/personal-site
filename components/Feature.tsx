import { Box } from "@chakra-ui/layout";
import { MotionBox } from "./MotionBox";
import SimpleBar from 'simplebar-react';
import { CloseButton } from "@chakra-ui/close-button";
import { useFeatureContext } from "../lib/featureContext";
import { AnimatePresence } from "framer-motion";

type FeatureProps = {
  project: number,
}

const containerVariants = {
  hidden: {
    y: "100vh"
  },
  visible: {
    y: "5vh",
    transition: {
      type: "spring",
      mass: 0.3,
      damping: 8,
      when: "beforeChildren",
      staggerChildren: 0.4,
    },
  },
  exit: {
    y: "100vh",
    transition: {
      type: "spring",
      mass: 0.3,
      damping: 8,
      duration: 5
    }
  }
}

const Feature = () => {
  const {projectToFeature, showFeature, closeFeature} = useFeatureContext();

  return (
    <AnimatePresence>
      { showFeature && (
        <>
          <Box position="absolute" top={0} right={0} bottom={0} left={0} bg="gray.100" />
      
          <MotionBox position="fixed" left={0} right={0} bottom={0} maxWidth="930px" marginY="0" marginX="auto" height="100vh"  variants={containerVariants} initial="hidden" animate="visible" exit="exit">
            
            <Box height="full" bg="white" borderWidth="1px" borderColor="gray.300" borderStyle="solid" borderRadius="xl">
              
              <SimpleBar style={{ maxHeight: '100%' }}>
                <CloseButton size="md" fontSize="16px" borderRadius="full" color="gray.900" bg="gray.100" p={0} top="10px" right="10px" onClick={closeFeature}  />
              </SimpleBar>
            </Box>
          </MotionBox> 
        </>
      )}
    </AnimatePresence>
  )
}

export default Feature;
