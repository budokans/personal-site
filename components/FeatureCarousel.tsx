import { Img } from "@chakra-ui/image";
import { Box, Stack } from "@chakra-ui/layout"
import { MotionBox } from "./MotionBox";

type FeatureCarouselProps = {
  media: string[]
}

const hideScrollbar = {
  '&::-webkit-scrollbar': {
    "width": "0 !important",
    "display": "none", 
  },
}

const containerVariants = {
  hidden: {
    x: "100vw",
  },
  visible: {
    x: 0,
    transition: {
      type: "spring",
      damping: 20,
    }
  }
}

const FeatureCarousel = ({media}: FeatureCarouselProps) => {
  return (
    <Box overflow="hidden" px={[4, 9]}>
      <MotionBox variants={containerVariants} initial="hidden" animate="visible" >

        <Stack 
          direction="row"
          alignItems="center"
          spacing={4} 
          height="calc(45vw + 35px)" 
          maxH="490px" 
          overflowX="scroll" 
          whiteSpace="nowrap" 
          sx={{ ...hideScrollbar, scrollSnapType: "x mandatory" }} 
          cursor="grab"
        >

        { media.map((image, idx) => {
          return (
            <Box 
              key={idx} 
              bg="gray.400" 
              borderRadius="xl"
              height="100%"
              minWidth="calc(70vw + 35px)"
              sx={{ scrollSnapAlign: "start", scrollPadding: "1.75rem" }} 
              p="7.5vw"
            >

              <Box h="100%" w="100%" >        
                <Img src={image} objectFit="cover" h="100%" w="100%" />
              </Box>
              
            </Box>
          )
        })}

        </Stack>
      </MotionBox>
    </Box>
  )
}

export default FeatureCarousel;