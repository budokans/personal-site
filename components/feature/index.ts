export * as Feature from "./Feature";
export * as Header from "./Header";
export * as Links from "./Links";
export * as Tech from "./Tech";
export * as Description from "./Description";
export { CarouselMobile } from "./CarouselMobile";
export { CarouselDesktop } from "./CarouselDesktop";
