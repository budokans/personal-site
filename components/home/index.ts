export * as Header from "./Header";
export * from "./Home";
export * as Portfolio from "./Portfolio";
