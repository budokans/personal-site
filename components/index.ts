export { DocHead } from "./DocHead";
export * as Feature from "./feature";
export * as Home from "./home";
export * as Link from "./Link";
export * as Motion from "./Motion";
