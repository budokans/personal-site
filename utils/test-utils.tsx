import { render, RenderOptions, RenderResult } from "@testing-library/react";
import { ChakraProvider } from "@chakra-ui/react";
import { ChildrenProps } from "types";
import { ReactElement } from "react";

const withProviders = ({ children }: ChildrenProps): ReactElement => {
  return <ChakraProvider>{children}</ChakraProvider>;
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "queries">
): RenderResult => render(ui, { wrapper: withProviders, ...options });

export * from "@testing-library/react";
export { customRender as render };
