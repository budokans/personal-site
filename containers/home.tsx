import { ReactElement } from "react";
import { SiteMetadata, Project } from "../types";
import { Header, Home, Portfolio } from "../components/home";
import { TextLink } from "../components/Link";

interface HomeContainerProps {
  readonly projects: readonly Project[];
  readonly metadata: SiteMetadata;
  readonly onPortfolioClick: (id: string) => void;
}

export const HomeContainer = ({
  projects,
  metadata,
  onPortfolioClick,
}: HomeContainerProps): ReactElement => {
  return (
    <Home.Home>
      <Header.Header>
        <Header.Heading>
          Hi, I&apos;m Steven. I&apos;m a full-stack developer who&apos;s helped
          build <TextLink url="https://rico.nz">Rico</TextLink> and{" "}
          <TextLink url="https://www.pivot-pointe.com">Pivot Pointe</TextLink>.
          I&apos;d love to help you with your software goals. Get in touch!
        </Header.Heading>

        <Header.HeaderContact>
          <Header.HeaderLocation>üìç {metadata.location}</Header.HeaderLocation>
          <Header.HeaderLinks contacts={metadata.contacts} />
        </Header.HeaderContact>
      </Header.Header>

      <Portfolio.Portfolio>
        {projects.map((project, idx) => {
          return (
            <Portfolio.PortfolioItem
              onPortfolioClick={onPortfolioClick}
              id={project.id}
              key={idx}
            >
              <Portfolio.PortfolioImage
                src={project.icon}
                alt={project.title}
              />
              <Portfolio.PortfolioInner
                idx={idx}
                projectsCount={projects.length}
              >
                <Portfolio.PortfolioTitle>
                  {project.title}
                </Portfolio.PortfolioTitle>
                <Portfolio.PortfolioText>
                  {project.shortBlurb}
                </Portfolio.PortfolioText>
              </Portfolio.PortfolioInner>
            </Portfolio.PortfolioItem>
          );
        })}
      </Portfolio.Portfolio>
    </Home.Home>
  );
};
