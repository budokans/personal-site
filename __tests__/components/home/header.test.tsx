import { render, within, screen, setDeviceWidth } from "test-utils";
import { Header } from "components/home";
import { testSiteMetadata } from "../../testData";

describe("<Header />", () => {
  test("renders <Header.Contact /> with data from props", () => {
    const { container } = render(
      <Header.Header>
        <Header.Location>{testSiteMetadata.location}</Header.Location>
        <Header.Links contacts={testSiteMetadata.contacts} />
      </Header.Header>
    );

    const location = screen.getByText(testSiteMetadata.location);
    expect(location).toBeInTheDocument();

    const listItems = screen.getAllByRole("listitem");
    expect(listItems).toHaveLength(testSiteMetadata.contacts.length);

    listItems.forEach((item, idx) => {
      const queries = within(item);
      const contact = testSiteMetadata.contacts[idx];

      switch (contact.type) {
        case "email":
          expect(queries.getByText("Email")).toBeInTheDocument();
          return;
        case "website":
          expect(queries.getByText(contact.name)).toBeInTheDocument();
          return;
      }
    });

    expect(container.firstChild).toMatchSnapshot();
  });

  test("renders <Header.TooltipBtn /> if viewport width >= 930px", () => {
    setDeviceWidth(930);

    const { container } = render(
      <Header.Links contacts={testSiteMetadata.contacts} />
    );

    expect(screen.getByTestId(/Tooltip/i)).toBeInTheDocument();
    expect(screen.queryByTestId(/mailto/i)).not.toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });

  test("renders <Header.Link /> instead of <Header.TooltipBtn /> for the email link if viewport width < 930px", () => {
    setDeviceWidth(929);

    const { container } = render(
      <Header.Links contacts={testSiteMetadata.contacts} />
    );

    expect(screen.queryByTestId(/Tooltip/i)).not.toBeInTheDocument();
    expect(screen.getByTestId(/mailto/i)).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });
});
