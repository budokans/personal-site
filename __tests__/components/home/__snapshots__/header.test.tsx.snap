// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Header /> renders <Header.Link /> instead of <Header.TooltipBtn /> for the email link if viewport width < 930px 1`] = `
<div
  class="chakra-wrap css-0"
>
  <ul
    class="chakra-wrap__list css-2mzr01"
  >
    <li
      class="chakra-wrap__listitem css-1yp4ln"
    >
      <a
        class="chakra-button css-acyrai"
        data-testid="mailto:contact@stevenwebster.co"
        href="mailto:contact@stevenwebster.co"
        rel="noopener noreferrer"
        target="_blank"
      >
        Email
      </a>
    </li>
    <li
      class="chakra-wrap__listitem css-1yp4ln"
    >
      <a
        class="chakra-button css-acyrai"
        data-testid="https://github.com/budokans"
        href="https://github.com/budokans"
        rel="noopener noreferrer"
        target="_blank"
      >
        GitHub
      </a>
    </li>
    <li
      class="chakra-wrap__listitem css-1yp4ln"
    >
      <a
        class="chakra-button css-acyrai"
        data-testid="https://www.linkedin.com/in/steven-webster-developer/"
        href="https://www.linkedin.com/in/steven-webster-developer/"
        rel="noopener noreferrer"
        target="_blank"
      >
        LinkedIn
      </a>
    </li>
  </ul>
</div>
`;

exports[`<Header /> renders <Header.Links /> with data from props 1`] = `
<div
  class="chakra-stack css-j99o02"
>
  <div
    class="chakra-wrap css-0"
  >
    <ul
      class="chakra-wrap__list css-2mzr01"
    >
      <li
        class="chakra-wrap__listitem css-1yp4ln"
      >
        <button
          class="chakra-button css-acyrai"
          data-testid="Tooltip"
          type="button"
        >
          Email
        </button>
      </li>
      <li
        class="chakra-wrap__listitem css-1yp4ln"
      >
        <a
          class="chakra-button css-acyrai"
          data-testid="https://github.com/budokans"
          href="https://github.com/budokans"
          rel="noopener noreferrer"
          target="_blank"
        >
          GitHub
        </a>
      </li>
      <li
        class="chakra-wrap__listitem css-1yp4ln"
      >
        <a
          class="chakra-button css-acyrai"
          data-testid="https://www.linkedin.com/in/steven-webster-developer/"
          href="https://www.linkedin.com/in/steven-webster-developer/"
          rel="noopener noreferrer"
          target="_blank"
        >
          LinkedIn
        </a>
      </li>
    </ul>
  </div>
</div>
`;

exports[`<Header /> renders <Header.Text /> with data from props 1`] = `
<div
  class="chakra-stack css-j99o02"
>
  <h1
    class="chakra-heading css-vsh8ld"
  >
    Hi, I'm Steve and I'm a full-stack developer who cares about scalability, performance, and intuitive UIs.
  </h1>
  <p
    class="chakra-text css-1x3yzs8"
  >
    Please use the links below to chat with me about how I can help build your project.
  </p>
</div>
`;

exports[`<Header /> renders <Header.TooltipBtn /> if viewport width >= 930px 1`] = `
<div
  class="chakra-wrap css-0"
>
  <ul
    class="chakra-wrap__list css-2mzr01"
  >
    <li
      class="chakra-wrap__listitem css-1yp4ln"
    >
      <button
        class="chakra-button css-acyrai"
        data-testid="Tooltip"
        type="button"
      >
        Email
      </button>
    </li>
    <li
      class="chakra-wrap__listitem css-1yp4ln"
    >
      <a
        class="chakra-button css-acyrai"
        data-testid="https://github.com/budokans"
        href="https://github.com/budokans"
        rel="noopener noreferrer"
        target="_blank"
      >
        GitHub
      </a>
    </li>
    <li
      class="chakra-wrap__listitem css-1yp4ln"
    >
      <a
        class="chakra-button css-acyrai"
        data-testid="https://www.linkedin.com/in/steven-webster-developer/"
        href="https://www.linkedin.com/in/steven-webster-developer/"
        rel="noopener noreferrer"
        target="_blank"
      >
        LinkedIn
      </a>
    </li>
  </ul>
</div>
`;
